opm/simulators/timestepping/gatherConvergenceReport.cpp

```diff
@@ -34,192 +37,317 @@ namespace
-        MPI_Pack(&type, 1, MPI_INT, buf.data(), buf.size(), &offset, mpi_communicator);
-        MPI_Pack(&phase, 1, MPI_INT, buf.data(), buf.size(), &offset, mpi_communicator);
-        MPI_Pack(&value, 1, MPI_DOUBLE, buf.data(), buf.size(), &offset, mpi_communicator);
+        auto pack = [&buf, &offset, mpi_communicator]
+            (const auto* ptr, const int size, const auto type)
+        {
+            MPI_Pack(ptr, size, type,
+                     buf.data(), buf.size(), &offset,
+                     mpi_communicator);
+        };
+
+        const int type = static_cast<int>(m.type());
+        const int phase = m.phase();
+        const double value = m.value();
+
+        pack(&type, 1, MPI_INT);
+        pack(&phase, 1, MPI_INT);
+        pack(&value, 1, MPI_DOUBLE);
     }
 
     void packWellFailure(const ConvergenceReport::WellFailure& f,
-                         std::vector<char>& buf,
-                         int& offset, MPI_Comm mpi_communicator)
+                         std::vector<char>& buf, int& offset,
+                         MPI_Comm mpi_communicator)
     {
-        int type = static_cast<int>(f.type());
-        int severity = static_cast<int>(f.severity());
-        int phase = f.phase();
-        MPI_Pack(&type, 1, MPI_INT, buf.data(), buf.size(), &offset, mpi_communicator);
-        MPI_Pack(&severity, 1, MPI_INT, buf.data(), buf.size(), &offset, mpi_communicator);
-        MPI_Pack(&phase, 1, MPI_INT, buf.data(), buf.size(), &offset, mpi_communicator);
-        int name_length = f.wellName().size() + 1; // Adding 1 for the null terminator.
-        MPI_Pack(&name_length, 1, MPI_INT, buf.data(), buf.size(), &offset, mpi_communicator);
-        MPI_Pack(const_cast<char*>(f.wellName().c_str()), name_length, MPI_CHAR, buf.data(), buf.size(), &offset, mpi_communicator);
+        auto pack = [&buf, &offset, mpi_communicator]
+            (const auto* ptr, const int size, const auto type)
+        {
+            MPI_Pack(ptr, size, type,
+                     buf.data(), buf.size(), &offset,
+                     mpi_communicator);
+        };
+
+        const int type = static_cast<int>(f.type());
+        const int severity = static_cast<int>(f.severity());
+        const int phase = f.phase();
+        pack(&type, 1, MPI_INT);
+        pack(&severity, 1, MPI_INT);
+        pack(&phase, 1, MPI_INT);
+
+        // Add one for the null terminator.
+        const int name_length = f.wellName().size() + 1;
+        pack(&name_length, 1, MPI_INT);
+        pack(f.wellName().c_str(), name_length, MPI_CHAR);
     }
 
     void packConvergenceReport(const ConvergenceReport& local_report,
-                               std::vector<char>& buf,
-                               int& offset, MPI_Comm mpi_communicator)
+                               std::vector<char>& buf, int& offset,
+                               MPI_Comm mpi_communicator)
     {
+        auto pack = [&buf, &offset, mpi_communicator]
```

Comment:
> I like the readability this lambda brings! It is also very repetitive, so I think it could improved a little. What about making a small helper class that is constructed using the buffer, offset and mpi_communicator, make pack() a member template of it, and pass the helper around instead of the trio of arguments?

I've pushed an update that uses this idea, but with a slightly different approach.  I've created a special purpose utility class&ndash;currently available only in `gatherConvergenceReport.cpp`&ndash;that plugs into the serialisation framework (i.e., using the [`Serializer<>`](https://github.com/OPM/opm-common/blob/8cc3939ad273ea2abda4bc31c1d6ac1d56ac69a0/opm/common/utility/Serializer.hpp#L90c) and [`Mpi::Packer`](https://github.com/OPM/opm-simulators/blob/c2e704c84f48b320a46e9466c6b0105deec98bc1/opm/simulators/utils/MPIPacker.hpp#L187) classes, along with the per-type `serializeOp()` member functions) and defers conversion between in-memory and byte stream representation to that framework.  The serialisation framework is for the most part designed to handle the case of broadcasting information from the I/O rank to all other ranks, but with a bit of a hack using `Serializer<>::m_buffer` I was able to get something that would effectively communicate all convergence reports from all ranks to all ranks.